python/ycm/youcompleteme.py

```diff
@@ -292,6 +297,127 @@ def OnCurrentIdentifierFinished( self ):
     SendEventNotificationAsync( 'CurrentIdentifierFinished' )
 
 
+  def OnCompleteDone( self ):
+    complete_done_actions = self.GetCompleteDoneHooks()
+    for action in complete_done_actions:
+      action(self)
+
+
+  def GetCompleteDoneHooks( self ):
+    filetypes = vimsupport.CurrentFiletypes()
+    for key, value in self._complete_done_hooks.iteritems():
+      if key in filetypes:
+        yield value
+
+
+  def GetCompletedCompletions( self ):
+    latest_completion_request = self.GetCurrentCompletionRequest()
+    if not latest_completion_request or not latest_completion_request.Done():
+      return []
+
+    completions = latest_completion_request.RawResponse()
+
+    result = self.FilterToCompletedCompletions( completions,
+                                                full_match_only = True )
+    result = list( result )
+    if result:
+      return result
+
+    if self.HasCompletionsThatCouldBeCompletedWithMoreText( completions ):
+      # Since the way that YCM works leads to CompleteDone called on every
+      # character, return blank if the completion might not be done. This won't
+      # match if the completion is ended with typing a non-keyword character.
+      return []
+
+    result = self.FilterToCompletedCompletions( completions,
+                                                full_match_only = False )
+
+    return list( result )
+
+
+  def FilterToCompletedCompletions( self, completions, full_match_only ):
```

Comment:
It's kinda hard to tell at-a-glance what this function does; the name might need to be clearer, and a doc comment would probably help.